### ISSUE TITLE
ORM should have a datetime with timezone type

### ISSUE THREAD

**Issue by tedicela** [cakephp/cakephp#13646]
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.8.1.
* Platform and Target: PHP 7.2.19, PostgreSQL

### What I did
I have to insert a record with a column datatype "TIMESTAMP WITH TIME ZONE". 
```
$data = [
    "title"=>"Test",
    "activity_start"=> "2019-09-04T00:00:00+02:00",
];

$activity = $this->Activities->newEntity();
$activity = $this->Activities->patchEntity($activity, $data);

var_dump($activity);

$this->Activities->save($activity);
```

The "var_dump" has printed:
```
object(App\Model\Entity\Activity)#6548 (16) {
  ["title"]=>
  string(7) "Test"
  ["activity_start"]=>
  object(Cake\I18n\FrozenTime)#6516 (3) {
    ["time"]=>
    string(25) "2019-09-04T00:00:00+02:00"
    ["timezone"]=>
    string(6) "+02:00"
    ["fixedNowTime"]=>
    bool(false)
  }
  ["[new]"]=>
  bool(true)
  ["[accessible]"]=>
  array(2) {
    ["*"]=>
    bool(true)
    ["id"]=>
    bool(false)
  }
  ["[dirty]"]=>
  array(7) {
    ["title"]=>
    ["activity_start"]=>
  }
  ["[original]"]=>
  array(0) {
  }
  ["[virtual]"]=>
  array(0) {
  }
  ["[hasErrors]"]=>
  bool(false)
  ["[errors]"]=>
  array(0) {
  }
  ["[invalid]"]=>
  array(0) {
  }
  ["[repository]"]=>
  string(10) "Activities"
}
```

### What happened
The record created on DB has lost the timezone:
```
id                   | 2138
activity_start    | 2019-09-04 00:00:00+00
```

### What I expected to happen
I was expecting to see on DB the datetime converted from +02:00 to UTC timezone:

```
id                   | 2138
activity_start    | 2019-09-03 22:00:00+00
```

*** This issue came after an upgrade from cakephp 3.3.16 to 3.8.1. ***

### After some investigations inside CakePHP I found out where this conversions happen. 
It happens on:
*file* - `Database/Type/DateTimeType.php` 
*method* - `toDatabase($value, Driver $driver)` 
*line* - 151 `return $value->format(array_shift($format));`

- The variable `$format` gets it's value on line 140:
 `$format = (array)$this->_format;` 

- `$this->_format` is defined on line 71:

```
protected $_format = [
	'Y-m-d H:i:s',
	'Y-m-d\TH:i:sP',
];
```

So returning to the line 151 using `array_shift($format)` will always use the format `Y-m-d H:i:s`. Using always this format means that it always will loose the timezone.

**Comment by markstory** [comment]
The datetime format you've located was last changed in 5bd3485a7d8 as part of 3.5.0. Prior to 3.5, the format string also did not include timezone information. Are you sure you were getting timezone data set before, as I don't think we've ever supported the timezone features of postgres.

**Comment by othercorey** [comment]
Are you saying you don't support simply passing the string to postgres and letting it do the conversion entirely on postgres-side?

**Comment by othercorey** [comment]
Maybe there is a way to tell CakePHP to convert to UTC before saving for all db queries to avoid having to manually force datetime object to convert to UTC before each db call - preferably without overriding the datetime type.

This would be a nice save coming from code that used 'timestamp with timezone' db code where the assumption (and correctly) is made that postgres converts the timestamp to utc before saving.

**Comment by ADmad** [comment]
> Maybe there is a way to tell CakePHP to convert to UTC before saving for all db queries to avoid having to manually force datetime object to convert to UTC before each db call - preferably without overriding the datetime type.

Changing timezone when converting datetime object to string when saving value to db is support in 4.x. You just have to configure the timezone use use use.
For e.g. add the following in your bootstrap `TypeFactory::build('timestamp')->setTimezone('UTC')`.

**Comment by othercorey** [comment]
I think for 3.8, you could subclass DateTimeType and override the _format array to make the format with timezone the first entry.

https://github.com/cakephp/cakephp/blob/a2ff473a20f467458680b6841a3fdb190eef113d/src/Database/Type/DateTimeType.php#L71-L74

This would let you insert/update with microseconds as well in postgresql.

**Comment by othercorey** [comment]
@markstory I don't want to hijack this thread, but I was curious why cakephp defaults to not formatting datetime with fractional seconds.  I am assuming this is legacy from older versions of mysql but even 5.5 supported timestamp strings with fractional seconds in it even though they discarded the fraction part.

**Comment by markstory** [comment]
@othercorey Mostly historical reasons. While MySQL supports fractional seconds it requires the `DATETIME(6)` column type which we haven't historically used. We could add new type classes for microsecond precision which would enable people to remap their datetime/timestamp types to more those implementations.

**Comment by othercorey** [comment]
I think the only difference is the $_format property I mentioned above.  Perhaps instead of requiring a whole new type, you could extend it in a similar manner to `useImmutable()` with something like `useFractionalSeconds()`.

The conversion from db to php simply relies on `DateTime::modify()`

**Comment by tedicela** [comment]
> The datetime format you've located was last changed in [5bd3485](https://github.com/cakephp/cakephp/commit/5bd3485a7d8e6e9ace109b47ac1e0805018f5d79) as part of 3.5.0. Prior to 3.5, the format string also did not include timezone information. Are you sure you were getting timezone data set before, as I don't think we've ever supported the timezone features of postgres.

Yes I'm sure CakePHP has supported ...maybe involuntary. Actually we have our webapp in CakePHP 3.3.16 that it works(currently in production) and the upgraded app with CakePHP 3.8.1 that it don't work.

I see that in the 3.3.16 it also converts DateTime into `Y-m-d H:i:s` but the difference is in the `patchEntity` method. Why?

on 3.316:
the patched $entity has Dates/DateTimes as strings (ex: what was received from response )

meanwhile on the 3.8.1:
the patched $entity has Dates/DateTimes as `Cake\I18n\FrozenTime` objects.

Personally I think the patchEntity actually is correct, but the conversion before saving into DB is not correct.

**Comment by markstory** [comment]
Sounds like the timezones were working by accident before. We certainly did not have any tests or documentation indicating that datetime with timezone types were supported before or now.

As for making it work in the future, there would need to be additional metadata present for datetime column types to allow timezone information to be conditionally included as including it when the columns don't support it will result in errors.

Removing the bug label on this as we've never officially supported datetimes include timezones, but could in the future.

If you'd like to restore the behavior where your columns are treated as strings instead of datetime objects you can use the [_initializeSchema](https://book.cakephp.org/3.0/en/orm/saving-data.html#saving-complex-types) method to replace the type with `string`.

**Comment by tedicela** [comment]
@markstory About how it can be implemented: I can tell you that if the database Driver is PostgreSQL converting always into the format `Y-m-d\TH:i:sP` will solve the problem in all cases of timestamp types because PostgreSQL inside can do casting if there is a string with valid datetime format. 

- If you give to postgresql a string with timezone but the column is "without time zone" it will drop the timezone.
- if you give to PostgreSQL a string without timezone but the column is "with time zone" it will assume that the timezone is the timezone of which the client has connected.
- when you `SELECT` from PostgreSQL the `timestamp with time zone` you will get it transformed in the timezone the client has specified in the connection.


I have not experimented timezones with other databases to give you more information.

**Comment by markstory** [comment]
I'm confident that postgres will be ok. However, I don't know that the same is true for sqlite, mysql and sqlserver. If you'd like you can put a pull request together with your intended changes. That will run our test suite against all the supported databases and we can see what happens.

**Comment by othercorey** [comment]
> Changing timezone when converting datetime object to string when saving value to db will be supported in Cake 4.0. You just have to configure the timezone to use.
> 
> For e.g. add the following in your bootstrap `TypeFactory::build('timestamp')->setTimezone('UTC')`.

@ADmad This seems to be supported in 3.8, no? I mean this requires the conversion is done on the client, but it should have the same result.

https://api.cakephp.org/3.8/class-Cake.Database.Type.DateTimeType.html#_setTimezone

```
        if ($this->dbTimezone !== null
            && $this->dbTimezone->getName() !== $value->getTimezone()->getName()
        ) {
            if (!$value instanceof DateTimeImmutable) {
                $value = clone $value;
            }
            $value = $value->setTimezone($this->dbTimezone);
        }

        return $value->format(array_shift($format));
```

**Comment by othercorey** [comment]
@markstory I did some testing as this is related to some features I'm interesting in.  I did this against 5.6 which is the oldest supported version I could use.  Note: 5.5 documentation states the same as 5.6 that it will accept fractional seconds in the string even though the type doesn't exist if 5.5 is important. I tried offset timezone information as well as timezone names similar to what postgres would accept. 

Here the `ts' and 'ts_micro' are TIMESTAMP and TIMESTAMP(6) types while 'dt' and 'dt_micro' are DATETIME and DATETIME(6) types.

This is with the default UTC server timezone.

```
mysql> INSERT INTO time_test (ts, dt, ts_micro, dt_micro) VALUES('2019-09-25 04:41:00-500', '2019-09-25 04:41:00-500', '2019-09-25 04:41:00-500', '2019-09-25 04:41:00-500');
Query OK, 1 row affected, 4 warnings (0.00 sec)

mysql> select * from time_test;
+---------------------+---------------------+----------------------------+----------------------------+
| ts                  | dt                  | ts_micro                   | dt_micro                   |
+---------------------+---------------------+----------------------------+----------------------------+
| 2019-09-25 04:41:00 | 2019-09-25 04:41:00 | 2019-09-25 04:41:00.000000 | 2019-09-25 04:41:00.000000 |
+---------------------+---------------------+----------------------------+----------------------------+
1 row in set (0.00 sec)

mysql> INSERT INTO time_test (ts, dt, ts_micro, dt_micro) VALUES('2019-09-25 04:41:00.123-500', '2019-09-25 04:41:00.123-500', '2019-09-25 04:41:00.123-500', '2019-09-25 04:41:00.123-500');
Query OK, 1 row affected, 4 warnings (0.00 sec)

mysql> select * from time_test;
+---------------------+---------------------+----------------------------+----------------------------+
| ts                  | dt                  | ts_micro                   | dt_micro                   |
+---------------------+---------------------+----------------------------+----------------------------+
| 2019-09-25 04:41:00 | 2019-09-25 04:41:00 | 2019-09-25 04:41:00.123000 | 2019-09-25 04:41:00.123000 |
+---------------------+---------------------+----------------------------+----------------------------+
1 row in set (0.00 sec)

mysql> INSERT INTO time_test (ts, dt, ts_micro, dt_micro) VALUES('2019-09-25 04:41:00.123 America/New_York', '2019-09-25 04:41:00.123 America/New_York', '2019-09-25 04:41:00.123 America/New_York', '2019-09-25 04:41:00.123 America/New_York');
Query OK, 1 row affected, 4 warnings (0.01 sec)

mysql> select * from time_test;
+---------------------+---------------------+----------------------------+----------------------------+
| ts                  | dt                  | ts_micro                   | dt_micro                   |
+---------------------+---------------------+----------------------------+----------------------------+
| 2019-09-25 04:41:00 | 2019-09-25 04:41:00 | 2019-09-25 04:41:00.123000 | 2019-09-25 04:41:00.123000 |
+---------------------+---------------------+----------------------------+----------------------------+
1 row in set (0.00 sec)
```

**Comment by markstory** [comment]
@othercorey That is good information. It is looking like including the timezone in the database parameters will be ok.

**Comment by ADmad** [comment]
I checked on MySQL 5.7 and trying to save `2007-03-18 15:30:23-500` for a `datetime` field throw error:
`Incorrect datetime value: '2007-03-18 15:30:23-500' for column 'created' at row`

**Comment by garas** [comment]
MySQL 5.5 doesn't support `TIMESTAMP(6)` and `DATETIME(6)` types.
On `TIMESTAMP` and `DATETIME` MySQL 5.5 accepts `'2019-09-25 04:41:00.123-500'`, but fails to parse `'2019-09-25 04:41:00.123 America/New_York'` and silently saves `0000-00-00 00:00:00` ⚠

MySQL 5.6 emits warnings if it truncates value (all queries above had `Query OK, 1 row affected, 4 warnings (0.01 sec)`).

MySQL 5.7/8.0 is strict by default and fails with error instead of warning. 

Sqlite saves and returns date strings as is.

**Comment by othercorey** [comment]
It seems like datetime and timezone strings really should be driver-specific or have a very simple on/off configuration switch.

I know CakePHP wants to have a simple one-size-fits-all solution, but this is one of the most diverse feature implementations in DBs.

The same with fractional seconds.  Even if we could assume all formats and all drivers support them somewhat equally, it would be nice to have it configured in the same way.

**Comment by othercorey** [comment]
@garas CakePHP says it supports SQLServer 2008.  I don't have access to that -- only 2017 through Docker. 

The default `datetime` column type does not support timezone offsets and the fractional seconds support requires rounding of the value which is not ideal.

However, it looks like the new `datetime2` and `datetimeoffset` types which do support proper fractional seconds and time zone offset strings were added in SQL Server 2008.  I cannot test against that version through -- so like mysql 5.5 - 8.0, I can't guarantee the result of the conversion with or without warnings.

SqlServerSchema does seem to support these types through:

```
        if (strpos($col, 'datetime') !== false) {
            return ['type' => TableSchema::TYPE_TIMESTAMP, 'length' => null];
        }
```

**Comment by tedicela** [comment]
I have did some tests in MySQL 5.5.59 and it seems that if give to MySQL a string with time zone it ignores the time zone. To handle time zones in MySQL should be used:

```
SET time_zone = '+03:00';
```

or the time zone can be saved in another column as a string.

So I guess using string with time zone will be ok with PostgreSql, MySql and SQLite. I don't know what can happen with Sql Server. I'm making a pull request then you can tests and accept/reject the PR.

**Comment by othercorey** [comment]
@tedicela Have you tried using the setTimezone() feature that @ADmad mentioned earlier and I noted was in 3.8?

https://github.com/cakephp/cakephp/issues/13646#issuecomment-534939211

**Comment by tedicela** [comment]
> @tedicela Have you tried using the setTimezone() feature that @ADmad mentioned earlier and I noted was in 3.8?
> 
> [#13646 (comment)](https://github.com/cakephp/cakephp/issues/13646#issuecomment-534939211)

I didn't tried because @ADmad mentioned that it will be available on CakePHP 4.0. 
Then using: 
```
$value->setTimezone(TIMEZONE);
```
currently will not make any difference since the time zone is always dropped and never provided in the queries that updates/inserts.  Anyway in the case I have explained above the datetime object already has the time zone(and there is no need to change the timezone):

```
 ["activity_start"]=>
  object(Cake\I18n\FrozenTime)#6516 (3) {
    ["time"]=>
    string(25) "2019-09-04T00:00:00+02:00"
    ["timezone"]=>
    string(6) "+02:00"
    ["fixedNowTime"]=>
    bool(false)
  }
```

I think that CakePHP 3 is  ok just to handle timezone like this(at least to give the time zone in the string), because all already implemented apps that uses CakePHP 3 has already implemented a way to handle time zones. In CakePHP 4 can implement a way to handle time zones through all Database drivers. Because for MySQL I guess to handle time zones current users of CakePHP 3 has used another column varchar just to save time zone or before they have to insert/update a timestamp with time zone they execute the query `SET time_zone = '+02:00' `.

**Comment by tedicela** [comment]
#13695

**Comment by othercorey** [comment]
@tedicela The setTimezone() option tells cakephp to convert a specific timezone when savings to db.  If you set it to the same timezone as your DB, then the missing timezone information shouldn't matter.

**Comment by othercorey** [comment]
@markstory I was able to add support for outputting fractional seconds and timezones to the db locally without destroying existing/legacy support.

However, trying to add support for storing the fractional seconds from the database string when converting to php failed.  By default, the DateTime parser doesn't recognize fractional second and timezone without the long format like '2008-07-01T22:35:17.03+08:00'.  We tend to see it like this: '008-07-01 22:35:17.03+08:00'.

This modify only supports these formats by default:
https://www.php.net/manual/en/datetime.formats.compound.php
```
    public function toPHP($value, Driver $driver)
    {
        ...
        $instance = clone $this->_datetimeInstance;
        $instance = $instance->modify($value);
        ...
        return $instance;
    }
```

I wasn't sure if the best approach would be a db driver specific format string to use with DateTime::createFromFormat() or to have yet another type-specific format similar to the one used by marshal() for locale-specific parsing.  It is unfortunate that microseconds aren't parsed in this common format.

```
-    'c' => 2001-01-04T12:13:14.315113+0000
-    'd' => 2001-01-04T12:13:14.315113+0200
+    'c' => 2001-01-04T12:13:14.000000+0000
+    'd' => 2001-01-04T12:13:14.000000+0000
```

**Comment by ADmad** [comment]
One thing to keep in mind is the same type class instance is used for all fields of that type. Now someone might have some fields with fractional seconds in their db and some without. So we should have a separate type class for support with fractional seconds.

**Comment by othercorey** [comment]
@ADmad After trying different approaches, I agree that's the cleanest approach especially without having to cleanup the api for 4.x.  One of the issue is these types all derive from DateTimeType which has some confusing code.  Adding more interface-based options like useFractional() and useTimezone() really didn't look good.

Example below.  There is an interface helper to set a locale parser format.  The documentation says it is used for parsing dates in this class, but it's only used for marshal() not toPHP().  Adding more format setters in the interface makes this extremely confusing to configure.

The marshal() function relies on comparing the marshaled type to the original value string.  This uses an array of DateTime format types in $_format via compare().  However, there's no real indication why _compare() needs these extra formats.  Clearly someone ran into an edge case somewhere when testing and dumped that extra format in to fix it.  This makes $_format extremely confusing because you don't know why it has an array of values where only of them is used in toDatabase() and neither is used in toPHP().

https://github.com/cakephp/cakephp/blob/a58519d7af33acbdb4914df7559e3fb0b2ea364f/src/Database/Type/DateTimeType.php#L71-L74

https://github.com/cakephp/cakephp/blob/a58519d7af33acbdb4914df7559e3fb0b2ea364f/src/Database/Type/DateTimeType.php#L248-L256

https://github.com/cakephp/cakephp/blob/a58519d7af33acbdb4914df7559e3fb0b2ea364f/src/Database/Type/DateTimeType.php#L299-L308

https://github.com/cakephp/cakephp/blob/a58519d7af33acbdb4914df7559e3fb0b2ea364f/src/Database/Type/DateTimeType.php#L334-L348

**Comment by ADmad** [comment]
>  There is an interface helper to set a locale parser format.

Personally I never liked that feature but I am sure there are people using it so we can't remove/deprecate it. I always show POST datetime values in format supported by database regardless of what format the date is displayed to user. Incidentally that's how HTML5's `datetime-local` input works, regardless of the display format the value is always POSTed in [yyyy-MM-ddThh:mm](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#Value) format. There are JS datepickers like "flatpickr" which also different formats for display and submitted values.

>  However, there's no real indication why _compare() needs these extra formats. Clearly someone ran into an edge case somewhere when testing and dumped that extra format in to fix it. This makes $_format extremely confusing because you don't know why it has an array of values where only of them is used in toDatabase() and neither is used in toPHP()

Yes a descriptive docblock would be useful there. The multiple formats are useful for marshalling only. It was a bit of a hack to allow submitting date time string in any format listed in that array and be marshalled into `Time` / `FrozenTime` object.

**Comment by markstory** [comment]
Fixed in #13943
